crates/analyzer-core/tests/lsp_position_struct.rs

```diff
@@ -0,0 +1,433 @@
+
+	// Boundary check on Range
+	let mut lsp = original_lsp.clone();
+	let event = change_event((7, 0), (7, 0), "xyz".into()); // line doesn't exist
+	lsp.lazy_add(&event);
+	let expected_lsp = LspPos::parse_file(&"012\n456\n\n9\nbcde\nxyz".to_string()); // add to end of file
+	assert_eq!(expected_lsp.get_ranges(), lsp.get_ranges());
+
+	let mut lsp = original_lsp.clone();
+	let event = change_event((1, 10), (1, 15), "xyz".into()); // character doesn't exist
+	lsp.lazy_add(&event);
+	let expected_lsp = LspPos::parse_file(&"012\n456xyz\n\n9\nbcde\n".to_string()); // add to end of line
+	assert_eq!(expected_lsp.get_ranges(), lsp.get_ranges());
+
+	let mut lsp = original_lsp.clone();
+	let event = change_event((1, 1), (1, 0), "xyz".into()); // Range end is smaller than start, produce panic
+	let res = std::panic::catch_unwind(move || lsp.lazy_add(&event)); // catch panic
+	assert!(res.is_err()); // make sure it paniced
+}
+
+#[test]
+fn exhaustive_lazy_add() {
+	// test strings
+	let test_files = [
+		"012\n456\n\n9\nbcde\n".to_string(), // end in '\n'
+		"\n012\n456\n\n9\nbcdef".to_string(),  // doesn't end in '\n'
+		// uncomment the long string for benchmarks
+//		"Alan\nTuring\nwas\nan\nEnglish\nmathematician,\nlogician,\nand\ncomputer\nscientist.\nHe\nis\nwidely\nconsidered\none\nof\nthe\nfounders\nof\nthe\nfield\nof\ntheoretical\ncomputer\nscience\nand\nartificial\nintelligence.\nDuring\nWorld\nWar\nII,\nTuring\nplayed\na\nvital\nrole\nin\nbreaking\nthe\nGerman\nEnigma\ncode,\nwhich\nhelped\nthe\nAllied\nforces\nin\ntheir\nfight\nagainst\nthe\nNazis.\nTuring's\nwork\non\ncomputability\nand\nthe\nconcept\nof\na\nTuring\nmachine\nlaid\nthe\nfoundation\nfor\nmodern\ncomputer\nscience.\nDespite\nhis\ncontributions,\nTuring\nfaced\npersecution\nfor\nhis\nhomosexuality\nand\nwas\nconvicted\nof\n\"gross\nindecency.\"\nHe\npassed\naway\nin\n1954,\nbut\nhis\nlegacy\nand\ncontributions\nto\nthe\nfield\nof\ncomputing\ncontinue\nto\nbe\ncelebrated\nto\nthis\nday.".to_string(),	// long example
+		"".to_string(),						 // empty file
+	]; 
+
+	let test_changes = ["", "x", "xy", "xyz", "\n", "\n\n", "\n\n\n", "\nx\n"];
+
+	//timers
+	let mut lazy_timer = Duration::new(0, 0);
+	let mut parse_timer = Duration::new(0, 0);
+
+	// exhaustive search
+	for file in test_files {
+		let original_lsp = LspPos::parse_file(&file.clone()); // Create default LspPos
+		for change in test_changes {
+			for size in 0..file.len() + 1 {
+				for start_byte in 0..(file.len() - size + 1) {
+					// generate Event
+					let start = original_lsp.byte_to_lsp(start_byte);
+					let end = original_lsp.byte_to_lsp(start_byte + size);
+
+					let (l1, c1) = (start.line, start.character);
+					let (l2, c2) = (end.line, end.character);
+					let event = change_event((l1, c1), (l2, c2), change.to_string());
+
+					let mut lsp = original_lsp.clone();
+
+					let clock = Instant::now();
+					lsp.lazy_add(&event);
+					lazy_timer += clock.elapsed();
```

Comment:
For smaller files, `lazy_add()` is actually slower than just re-parsing the file. From tests, a file with 60 characters and 12 '\n' occurrences is the turning point for when `lazy_add()` is more efficient